plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,270 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Authors: Răzvan Vîrtan <virtanrazvan@gmail.com>
+ *          Justin He     <justin.he@arm.com>
+ *
+ * Copyright (c) 2021, Arm Ltd., University Politehnica of Bucharest. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <uk/config.h>
+#include <libfdt.h>
+#include <uk/plat/common/sections.h>
+#include <kvm/console.h>
+#include <kvm/config.h>
+#include <uk/assert.h>
+#include <kvm-arm/mm.h>
+#include <kvm-arm/smp.h>
+#include <kvm/intctrl.h>
+#include <arm/cpu.h>
+#include <uk/arch/limits.h>
+#include <stdbool.h>
+#include <uk/plat/bootstrap.h>
+#include <uk/sched.h>
+#include <uk/plat/time.h>
+
+#include <uk/plat/memory.h>
+#include <uk/plat/io.h>
+#include <uk/plat/lcpu.h>
+#include <arm/psci.h>
+#include <ofw/fdt.h>
+
+static __u64 smp_numcores;
+__u64 bspdone;
+__u64 smp_aps_started; // APs that are started and reached their entry function
+__u64 smp_aps_running; // APs that are started and aware of bspdone = 1
+
+struct arm64_cpu {
+    uint64_t id;
+	ukplat_lcpu_entry_t entry;
+    void *stackp;
+    struct uk_list_head fnlist;
+};
+const size_t arm64_cpu_size = sizeof(struct arm64_cpu);
+
+struct arm64_cpu cpus[CONFIG_MAX_CPUS];
+
+static __lcpuid bspid;
+
+static void ndelay(__u64 nsec)
```

Comment:
I think we must have such delay functions in timer subsystem (and export them) instead of define them here as static. Eventually other driver(s) will do exactly same thing in order part(s) of the code, duplicating them....

Reply:
Yes, agree. We have similar `delay` functions in x86. In that case, they are based on just executing an `out` instruction and expecting it to take a certain amount of time. So it is rather unprecise but faster than doing a precise timer-based wait and good enough for the purpose. It also does not rely on the timer being up and running. We have to decide if we keep this and introduce an equivalent function on arm or if we toss it altogether.